### ISSUE TITLE
Chakra UI: `dangerouslySetInnerHTML` did not match.

### ISSUE THREAD

**Issue by MariuzM** [hashicorp/next-mdx-remote#72]
When using components from Chakra UI i get this error: 

```
react-dom.development.js?61bb:67 Warning: Prop `dangerouslySetInnerHTML` did not match. Server: "<div>Some **mdx** text, with a component <p class=\"chakra-text css-bk9fzy\">Test</p></div>" Client: "<div>Some **mdx** text, with a component <style data-emotion=\"css bk9fzy\">.css-bk9fzy{font-size:20px;}</style><p class=\"chakra-text css-bk9fzy\">Test</p></div>"
```

```js
import { Text } from '@chakra-ui/react'
import hydrate from 'next-mdx-remote/hydrate'
import renderToString from 'next-mdx-remote/render-to-string'

const components = { Text }

export default function TestPage({ source }: any) {
  const content = hydrate(source, { components })
  return <div className="wrapper">{content}</div>
}

export async function getStaticProps() {
  const source = '<div>Some **mdx** text, with a component <Text fontSize={20}>Test</Text></div>'
  const mdxSource = await renderToString(source, { components })
  return { props: { source: mdxSource } }
}

```

For now the solution just remove the `components` from staticProps, like so and it works, bus is there a better solution?:

```js
export async function getStaticProps() {
  const source = '<div>Some **mdx** text, with a component <Text fontSize={20}>Test</Text></div>'
  const mdxSource = await renderToString(source)
  return { props: { source: mdxSource } }
}
```

**Comment by jescalan** [comment]
If you could dig a little deeper and make a minimal reproduction, ideally without Chakra UI, that would put me in a much better place to be able to triage this issue - otherwise I'm not sure I can get this into my own priorities as we do not use Chakra UI on my team.

**Comment by MariuzM** [comment]
Without Chakra UI it works fine, not sure what could i look for, it's a standart Next.js install with Chakra UI on top
But in nutshell these are the difference hence the error: 

```
Server: "<div>Some **mdx** text, with a component <p class=\"chakra-text css-bk9fzy\">Test</p></div>" 
Client: "<div>Some **mdx** text, with a component <style data-emotion=\"css bk9fzy\">.css-bk9fzy{font-size:20px;}</style><p class=\"chakra-text css-bk9fzy\">Test</p></div>"
```

I can only assume that static prop is not rendering maybe the component fully before passing to client

**Comment by jescalan** [comment]
Yeah I am really sorry but I don't think I can help with this. If you are able to get closer to the root of the problem, or if someone else sees this issue and is willing to put some time into helping you out, those are the roads I can see to getting closer to an answer!

**Comment by dayvista** [comment]
I am getting the same error with versions 1.0.1 of Chakra UI and 10.0.3 of Next.js. I don't have time to delve deeper at the moment, just wanted to mention it. The hydrated JSX containing the Chakra components still renders exactly as intended, despite this rather intimidating error.

**Comment by MariuzM** [comment]
You can get rid of that error by removing component from SSR but then you will just get this warning in SSR 
`Component Box was not imported, exported, or provided by MDXProvider as global scope`

**Comment by dayvista** [comment]
I believe the issue is that the string I'm passing into next-mdx-remote includes custom components that render into basic HTML tags when hydrated client side. So what ultimately gets displayed client side (a <p></p> tag for instance) ends up being different from what was passed into the renderToString function (a <Box/> component from Chakra UI).

**Comment by rodbs** [comment]
Same issue with next 10.0.4 and @chaka-ui/react 1.0.4

The funny thing is this error goes away when I import the NavBar in _app.js (built with Chakra-Ui), dinamically. I don't know why.
```
const LayoutDynamic = dynamic(() => import('@/components/Layout'));
```
```
  <LayoutDynamic activeTab={activeTab} setActiveTab={setActiveTab}>
               <Component {...pageProps} />
         
        </LayoutDynamic>
```

**Comment by igeligel** [comment]
I have the same issue and it might be even breaking my production environment. My repository is open-source, it's not a minimal reproduction but will work on it tomorrow.

Relevant code can be found here: https://github.com/igeligel/personal-site/blob/master/pages/%5Bpost%5D.tsx#L67-L192

Quick installation instructions for the repository: `yarn && yarn dev` -> Open http://localhost:3000/composing-computed-properties-in-vue-js for example.

On production, it looks even weirder: https://www.kevinpeters.net/composing-computed-properties-in-vue-js Opening that results in broken styles for some reason. When changing the route in the app by clicking on `Articles` on top and getting back to any article will make the styles work. Hence, in production, I am statically exporting the pages. Might be a reason why, but when running locally I could not reproduce that (with `yarn build && yarn export && npx http-server ./out`)

**Comment by jescalan** [comment]
If you can get this down to a minimal reproduction I'd be happy to take a look. With source code like that, the first step to debugging is cutting it down to a minimal repro regardless.

**Comment by igeligel** [comment]
Hey, here is a reproduction: https://github.com/igeligel/next-mdx-remote-72

So what I figured out. It breaks once you put in the responsive styles. The affected component is here: https://github.com/igeligel/next-mdx-remote-72/blob/main/components/ChakraParagraph.tsx#L5

Feel free to check it out. I have the feeling it has something to do with babel's compilation, but I will check later.

**Comment by MaximeHeckel** [comment]
I have the same issue but not with ChakraUI, but with Emotion (CSS-in-JS library) maybe that can help? I use next-mdx-remote for my blog posts:
- I see the error reported above on dev
- In prod though, I have a weird behavior: I can navigate to a post without a problem from my landing page (it doesn't use MDX), reloading a post though would cause the whole page to "block": I can see components being rendered, but style is missing, animations do not trigger, client side fetching will not trigger


Will try to do a minimal repro!

**Comment by jescalan** [comment]
Thanks so much to everyone for providing the reproductions here! We'll take a look at these soon and hopefully have something positive to report back 😀

**Comment by jescalan** [comment]
I pulled this down to take a look and didn't see anything that immediately stuck out to me unfortunately. We'll have to return a little later for a deeper dive, I think.

**Comment by MaximeHeckel** [comment]
Thank you for checking @jescalan! It really only happens with `next dev`  as far as I've checked so far, there are no side-effects or issues related to this on the build output

**Comment by miraklasiaf** [comment]
I have this problem too when using Chakra UI's component. This error caused FUOC in my website.

**Comment by miraklasiaf** [comment]
@jescalan my bad, I'm sorry. I've already solved the FOUC by adding the wrapper object within markdown component. This solution https://github.com/hashicorp/next-mdx-remote/issues/32#issuecomment-734016132 solves the issue for me. Feel free to check it out https://github.com/miraklasiaf/chakra-blog-starter. Thank you for your response!

**Comment by mohitgarg** [comment]
Hi @jescalan, I don't know how helpful this will be in resolving this issue but here's what I have found:

1. While passing the `MDXComponents` if there are any custom styles applied to the component, that's when the error happens. If you inspect the `renderedOutput` there's a `style `tag with some css being added to the HTML which I think is causing this issue.  If I remove any of the style props, there's no `dangerouslySetInnerHTML` error in the console
`<style data-emotion="css w0pt8a">.css-w0pt8a{margin-top:4px;margin-bottom:4px;color:red;}</style>`

2. Not passing custom components in `renderToString` doesn't throw the warning. Although it does cause the FOUC issue even if you passes the ThemeProvider as a wrapper

Here's a sandbox link
https://codesandbox.io/s/fervent-payne-ius37?file=/pages/day/index.js

**Comment by payapula** [comment]
Hi @jescalan , From what I analysed, I think this is related to emotion's SSR rendering. 

As Chakra uses emotion, we are facing this issue. Emotion have this [in the docs](https://emotion.sh/docs/ssr) mentioning about a possible SSR issue.

> Warning: This approach can interfere with nth child and similar selectors as it is inserts style tags directly into your markup. You will get a warning if you use such selectors when using this approach.

They also proposing a [advanced approach](https://emotion.sh/docs/ssr#advanced-approach) to overcome such warnings.

In my use case I get this problem only when reloading the MDX page. Removing the components from SSR removes the warning though I am not sure if this is recommended solution.

**Comment by BRKalow** [comment]
Thanks for the report and thanks for testing out v3 @payapula! I believe that this issue has been resolved per your comment on #121. Feel free to reference the upgrade guide in the [release notes](https://github.com/hashicorp/next-mdx-remote/releases/tag/3.0.0) to get onto latest. 🙏
